flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -4049,6 +4066,36 @@ class ArrayExprLowering {
     }
   }
 
+  /// Lower an elemental function array argument with a given
+  /// ConstituentSemantics.
+  template <typename A>
+  CC genElementalArgumentAs(const A &x, ConstituentSemantics semantics) {
+    // Ensure the returned element is in memory if this is what was requested.
+    if ((semantics == ConstituentSemantics::RefOpaque ||
+         semantics == ConstituentSemantics::DataAddr ||
+         semantics == ConstituentSemantics::ByValueArg)) {
+      if (!Fortran::evaluate::IsVariable(x)) {
+        PushSemantics(ConstituentSemantics::DataValue);
+        CC cc = genarr(x);
+        mlir::Location loc = getLoc();
+        if (isParenthesizedVariable(x)) {
+          // Parenthesised variables are lowered to a reference to the variable
+          // storage. When passing it as an argument, a copy must be passed.
+          return [=](IterSpace iters) -> ExtValue {
+            return createInMemoryScalarCopy(builder, loc, cc(iters));
+          };
+        }
+        mlir::Type storageType =
+            fir::unwrapSequenceType(converter.genType(toEvExpr(x)));
+        return [=](IterSpace iters) -> ExtValue {
+          return placeScalarValueInMemory(builder, loc, cc(iters), storageType);
```

Comment:
This case and line 4085 seems tauntingly similar. The code is ever so slightly different, but I wonder if it can't be common.

I think that `(VAR)` and `VAR` as a "data value" should produce the same result. There is a bit of nuance with a memory-bound object, I suppose though. Are we ever not passing a copy though and passing a reference directly to an array object's subelement?

Reply:
> I think that (VAR) and VAR as a "data value" should produce the same result. There is a bit of nuance with a memory-bound object, I suppose though.

`genarr()` of `(VAR)` and `VAR` with DataValue indeed produce the same thing (with the exception of the fir.no_reassoc), the issue is precisely with characters/derived that are handled in memory and for which the result is a reference to an element address in VAR. With `(VAR)`, we want to ensure we are passing the address of an element copy.
`placeScalarValueInMemory` does not do that because this would otherwise create needless extra temps when passing expression like `"prefix" // c1` where the address we have is already a temp. Hence the need for `createInMemoryScalarCopy`.

| Are we ever not passing a copy though and passing a reference directly to an array object's subelement

That is what would happen if skipping the `isParenthesizedVariable` case and lowering:
```
interface
elemental subroutine foo(c)
  character(10), value :: c
  ! ...
end subroutine
end interface

character :: c(10)
call foo((c))
```